practices/appeng/adrs/0006-use-graphql.md

```diff
@@ -0,0 +1,83 @@
+# 0006 - Use GraphQL for the API Transport Protocol
+
+**Status:** Draft by: MacRae
+
+**Reviewers:**
+
+## Context
+
+Most of our projects involve a web-app communicating with an API over HTTPS in some format or another. For most of my career, the default choice for what they send back and forth has been “REST”. For just as long, teams have had arguments about what exactly is the RESTful way to do things. GraphQL is a protocol that exposes your api as a graph of connected entities, allowing the client to fetch related entities in a single request. More importantly, it standardizes much of the things that robust REST implementations eventually implement on their own, leading to more robust generated code, and automatic clients being available in different languages.
+
+Specifically, with respect to it being a graph, it is very common for mature REST implementations to add features for including related objects in a request for a single object. You might `GET /customers/3?include=[‘address’, ‘shipments.approver’]` for instance, to reduce the number of requests that a client needs to make to render a specific page. That kind of thing is something every team that implements a REST API ends up needing to specify and implement themselves. GraphQL took that, and made it a core part of the protocol. If you have a page that needs those related entities, you just specify that in your request, and the server will return all of them together.
```

Comment:
From Suz: mention that it's for client/server by default, and also that it doesn't have to be your whole API